src/handlers/miner_hbbft_handler.erl

```diff
@@ -457,77 +478,189 @@ restore(OldState, NewState) ->
 %% helper functions
 fixup_msgs(Msgs) ->
     lists:map(fun({unicast, J, NextMsg}) ->
-                      {unicast, J+1, t2b(NextMsg)};
+                      {unicast, J+1, msg_to_bin(NextMsg)};
                  ({multicast, NextMsg}) ->
-                      {multicast, t2b(NextMsg)}
+                      {multicast, msg_to_bin(NextMsg)}
               end, Msgs).
 
-dedup_signatures(InSigs, #state{signatures = Sigs}) ->
-    %% favor existing sigs, in case they differ, but don't revalidate
-    %% at this point
-    lists:usort(Sigs ++ InSigs).
-
-enough_signatures(#state{artifact=undefined}) ->
-    false;
-enough_signatures(#state{sig_phase = sig, signatures = Sigs, f = F}) when length(Sigs) < F + 1 ->
-    false;
-enough_signatures(#state{sig_phase = done, signatures = Signatures}) ->
-    {ok, done, Signatures};
-enough_signatures(#state{sig_phase = Phase, artifact = Artifact, members = Members,
-                         f = F, signatures = Signatures, signatures_required = Threshold0}) ->
-    Threshold = case Phase of
-                    sig -> F + 1;
-                    gossip -> Threshold0
-                end,
+-spec sigs_to_positions([{A, _, _}], [A]) -> [pos_integer()].
+sigs_to_positions(Addr2Sig, Addresses) ->
+    Addr2Pos = lists:zip(Addresses, lists:seq(1, length(Addresses))),
+    Positions = lists:map(fun({Addr, _S, _V}) ->
+                            %% find member index
+                            {_, Pos} = lists:keyfind(Addr, 1, Addr2Pos),
+                            Pos
+                    end,
+                    Addr2Sig),
+    lists:usort(Positions).
+
+md_version(Ledger) ->
+    case blockchain:config(?election_version, Ledger) of
+        {ok, N} when N >= 3 ->
+            map;
+        _ ->
+            tuple
+    end.
+
+-spec state_reset(hbbft:hbbft_data(), #state{}) -> #state{}.
+state_reset(HBBFT, #state{}=S) ->
+    S#state{
+        hbbft          = HBBFT,
+        sigs_valid     = [],
+        sigs_invalid   = [],
+        sigs_unchecked = [],
+        artifact       = undefined,
+        sig_phase      = unsent,
+        bba            = <<>>,
+        seen           = #{}
+    }.
+
+-spec state_sigs_input(sigs(), #state{}) ->
+    {
+        {done, sigs()} | {gossip, sigs()} | pending,
+        #state{}
+    }.
+state_sigs_input(SigsIn, #state{hbbft=HBBFT}=S0) ->
+    CurRound = hbbft:round(HBBFT),
+    S1 = state_sigs_add(SigsIn, S0),
+    S2 = state_sigs_retry_unchecked(S1),
+    case state_sigs_check_if_enough(S2) of
+        {{some, {gossip, _}=Gossip}, S3} ->
+            {Gossip, S3#state{sig_phase = gossip}};
+        %% XXX What if artifact is undefined?
```

Comment:
You can't check the signatures without an artifact?

Reply:
Below, `Art` is passed into `ok = miner:signed_block(Sigs, Art),`, which would crash on `blockchain_block:deserialize(BinBlock)` if I'm reading the call chain correctly.